[iOS] Find on page highlight is broken for -webkit-user-select:none styled elements (affects many sites, including nytimes.com, bizjournal.com, cnn.com, developer.mozilla.org)

megangardner opened the pull request:
#### 47db418a5ff00139a0068296c759b0c776ef904a
<pre>
[iOS] Find on page highlight is broken for -webkit-user-select:none styled elements (affects many sites, including nytimes.com, bizjournal.com, cnn.com, developer.mozilla.org)
<a href="https://bugs.webkit.org/show_bug.cgi?id=297441">https://bugs.webkit.org/show_bug.cgi?id=297441</a>
<a href="https://rdar.apple.com/8268929">rdar://8268929</a>

Reviewed by Aditya Keerthi.

When we try and make a text indicator for a selection that was a
range that is webkit-user-select-none then the selection collapses
and the indicator does not have an associated image.

We should fallback to creating an indicator of the range painted
without the alterations for the indicator, for now. This will allow us
to at least have somewhat of an indicator, even if it is not as obvious.

We also shouldn&apos;t block on the existence of a selection rect for painting,
as if we have text rects and an image, that is also workable for showing
an indicator.

* Source/WebCore/page/TextIndicator.cpp:
(WebCore::TextIndicator::createWithRange):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::drawRect):

Canonical link: <a href="https://commits.webkit.org/298785@main">https://commits.webkit.org/298785@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1517e21592de1fdd948c8e1928889100377d066e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116640 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36304 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26866 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122714 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67212 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118529 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37002 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44893 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88585 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43002 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119589 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29511 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104645 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69051 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28575 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22744 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66381 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98891 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22902 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125850 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43539 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/32707 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97253 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43903 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100839 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97046 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24717 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42370 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20298 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39501 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43425 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49020 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42891 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46231 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44597 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -367,7 +367,7 @@ void WebFoundTextRangeController::drawRect(WebCore::PageOverlay&, WebCore::Graph
         graphicsContext.fillPath(path);
 
-    if (m_textIndicator && !m_textIndicator->selectionRectInRootViewCoordinates().isEmpty()) {
+    if (m_textIndicator && (!m_textIndicator->selectionRectInRootViewCoordinates().isEmpty() || !m_textIndicator->textBoundingRectInRootViewCoordinates().isEmpty())) {
```
I don't understand the existing check, and/or am not sure why we're not just replacing it? Nothing else in this block consults the selection rect...

megangardner replied:
Check was added in https://commits.webkit.org/259857@main to mitigate this issue in a different way. This added code makes it unnecessary

hortont424 commented on Source/WebCore/page/TextIndicator.cpp:
```diff
@@ -106,6 +106,9 @@ RefPtr<TextIndicator> TextIndicator::createWithRange(const SimpleRange& range, O
 #endif
     TemporarySelectionChange selectionChange(*document, { rangeToUse }, temporarySelectionOptions);
 
+    if (rangeToUse != document->selection().selection().toNormalizedRange())
```
This highly mysterious logic could use a comment! ("if the selection couldn't be set (perhaps due to user-select), we can't do a selection-only paint, so paint everything!")

pxlcoder commented on Source/WebCore/page/TextIndicator.cpp:
```diff
@@ -106,6 +106,10 @@ RefPtr<TextIndicator> TextIndicator::createWithRange(const SimpleRange& range, O
 #endif
     TemporarySelectionChange selectionChange(*document, { rangeToUse }, temporarySelectionOptions);
 
+    // if the selection couldn't be set (usually due to user-select), we can't do a selection-only paint, so changed to painting everything.
```
```suggestion
    // If the selection couldn't be set (usually due to user-select), we can't do a selection-only paint, so changed to painting everything.
```

megangardner committed:
[iOS] Find on page highlight is broken for -webkit-user-select:none styled elements (affects many sites, including nytimes.com, bizjournal.com, cnn.com, developer.mozilla.org)
https://bugs.webkit.org/show_bug.cgi?id=297441
rdar://8268929

Reviewed by Aditya Keerthi.

When we try and make a text indicator for a selection that was a
range that is webkit-user-select-none then the selection collapses
and the indicator does not have an associated image.

We should fallback to creating an indicator of the range painted
without the alterations for the indicator, for now. This will allow us
to at least have somewhat of an indicator, even if it is not as obvious.

We also shouldn't block on the existence of a selection rect for painting,
as if we have text rects and an image, that is also workable for showing
an indicator.

* Source/WebCore/page/TextIndicator.cpp:
(WebCore::TextIndicator::createWithRange):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::drawRect):

Canonical link: https://commits.webkit.org/298785@main
